### ISSUE TITLE
Spark Job Server on Mesos cluster: Unable to connect to server

### ISSUE THREAD

**Issue by DatPhanTien** [spark-jobserver/spark-jobserver#1097]
**Spark 2.3.1**

**Spark JobServer 0.8.0**

**Mesos 1.5.0**

**Spark cluster mode**

Dear,

We are using Spark Job Server for run our Spark applications on Mesos cluster.
Our cluster has Mesos and Spark running already. There is one Spark master node and 3 Spark workers. We start the Mesos-dispatcher server on the Spark master (listens on port 7077).
We successfully connect Spark Job Server to the Spark Rest API and run our apps with the following configuration: 
```
spark {
  master = "spark://spark_master:6066"
  submit.deployMode = "cluster"
  job-number-cpus = 4
  jobserver {
    port = 8090
    jobdao = spark.jobserver.io.JobSqlDAO
    context-per-jvm = true
    driver-mode = "mesos-cluster"
    sqldao {
      slick-driver = slick.driver.H2Driver
      # JDBC driver, full classpath
      jdbc-driver = org.h2.Driver
      # Directory where default H2 driver stores its data. Only needed for H2.
      rootdir = /tmp/spark-jobserver/sqldao/data
      # h2s
      jdbc {
        url = "jdbc:h2:tcp://spark_master:8083//tmp/spark-jobserver;AUTO_RECONNECT=TRUE"
        user = "secret"
        password = "secret"
      }
      dbcp {
        # enabled = false
        maxactive = 20
        maxidle = 10
        initialsize = 10
      }
      startH2Server = false
    }
  }
}
```
However, the running apps did not seem to be handled by the Mesos. This is actually strange since we could directly run our apps by POSTing to the Spark REST API (port 6066) and the execution is managed by Mesos normally.

We suspect that it was because we did not connect the Spark Job Server to the Mesos dispatcher.
However, we did not manage to connect the Spark Job server to the Mesos dispatcher (we did so by setting **master = "mesos://spark_master:7077"**). The error that we got when trying to create a new context was:
`Exception in thread "main" org.apache.spark.deploy.rest.SubmitRestConnectionException: Unable to connect to server
	at org.apache.spark.deploy.rest.RestSubmissionClient$$anonfun$createSubmission$3.apply(RestSubmissionClient.scala:104)
	at org.apache.spark.deploy.rest.RestSubmissionClient$$anonfun$createSubmission$3.apply(RestSubmissionClient.scala:86)
	at scala.collection.Traversabl2018-07-24 14:56:47 INFO  ShutdownHookManager:54 - Shutdown hook called
2018-07-24 14:56:47 INFO  ShutdownHookManager:54 - Deleting directory /tmp/spark-ed6bfff0-d580-446e-bc11-25e81e2708f9
s$ofRef.foreach(ArrayOps.scala:186)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
	at org.apache.spark.deploy.rest.RestSubmissionClient.createSubmission(RestSubmissionClient.scala:86)
	at org.apache.spark.deploy.rest.RestSubmissionClientApp.run(RestSubmissionClient.scala:429)
	at org.apache.spark.deploy.rest.RestSubmissionClientApp.start(RestSubmissionClient.scala:441)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:894)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:198)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:228)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:137)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: org.apache.spark.deploy.rest.SubmitRestConnectionException: No response from server
	at org.apache.spark.deploy.rest.RestSubmissionClient.readResponse(RestSubmissionClient.scala:271)
	at org.apache.spark.deploy.rest.RestSubmissionClient.org$apache$spark$deploy$rest$RestSubmissionClient$$postJson(RestSubmissionClient.scala:225)
	at org.apache.spark.deploy.rest.RestSubmissionClient$$anonfun$createSubmission$3.apply(RestSubmissionClient.scala:90)
	... 13 more
Caused by: java.util.concurrent.TimeoutException: Futures timed out after [10 seconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
	at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:190)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.Await$.result(package.scala:190)
	at org.apache.spark.deploy.rest.RestSubmissionClient.readResponse(RestSubmissionClient.scala:264)
	... 15 more`

Could you please take a look at let us know what did go wrong?!

Best
Tien Dat

**Comment by bsikander** [comment]
Have you tried putting the port `6066` in jobserver config instead of `7077`?

**Comment by DatPhanTien** [comment]
Well, perfect.
It worked with port 6066. To be precise, the config is:
`master = "mesos://spark_master:6066"`
Just wonder why this can be recognized? Since 6066 is the default REST API of Spark, does Spark job server do some sort of mapping for this configuration?

Just let me test whether jobs committed to this newly created context will be monitored by Mesos or not. 
I will let you know once I finish this test.

**Comment by DatPhanTien** [comment]
The application ran successfully. However, it did not appear to be launched by Mesos.

Then I tried to connect directly to the Mesos dispatcher server (which is open on port 7078, since 7077 is the default port for Spark master). The Spark job server managed to connect. 
However, I noticed that there was an issue with Spark Job Server context on Mesos cluster.
The error seems to happen at a low-level compatibility issue:
`I0724 16:41:57.305243     3 exec.cpp:162] Version: 1.5.0
I0724 16:41:57.311048    10 exec.cpp:236] Executor registered on agent a236b99a-e8f4-4b8e-9a26-56148ce4f55d-S0
I0724 16:41:57.312507     8 executor.cpp:171] Received SUBSCRIBED event
I0724 16:41:57.312912     8 executor.cpp:175] Subscribed executor on 10.82.101.157
I0724 16:41:57.316498     6 executor.cpp:171] Received LAUNCH event
I0724 16:41:57.317487     6 executor.cpp:638] Starting task driver-20180724164156-0001
I0724 16:41:57.324362     6 executor.cpp:478] Running '/opt/mesosphere/active/mesos/libexec/mesos/mesos-containerizer launch <POSSIBLY-SENSITIVE-DATA>'
I0724 16:41:57.326396     6 executor.cpp:651] Forked command at 13
Exception in thread "main" java.lang.IllegalArgumentException: Missing application resource.
	at org.apache.spark.launcher.CommandBuilderUtils.checkArgument(CommandBuilderUtils.java:241)
	at org.apache.spark.launcher.SparkSubmitCommandBuilder.buildSparkSubmitArgs(SparkSubmitCommandBuilder.java:160)
	at org.apache.spark.launcher.SparkSubmitCommandBuilder.buildSparkSubmitCommand(SparkSubmitCommandBuilder.java:274)
	at org.apache.spark.launcher.SparkSubmitCommandBuilder.buildCommand(SparkSubmitCommandBuilder.java:151)
	at org.apache.spark.launcher.Main.main(Main.java:86)
sh: line 1: -verbose:gc: command not found
sh: line 2: -XX:MaxPermSize=512m: command not found
sh: line 3: -XX:+CMSClassUnloadingEnabled: command not found
sh: line 4: -XX:+HeapDumpOnOutOfMemoryError: command not found
I0724 16:41:57.823385     7 executor.cpp:938] Command exited with status 127 (pid: 13)`

Do you have any experience with such behavior?

**Comment by velvia** [comment]
Hmmm it seems that there is some weird shell interaction with execution script. Which version are you using? Perhaps it is trying to start manager_start.sh, try putting some extra debugging in there.

**Comment by DatPhanTien** [comment]
As I put in the first command:
Spark 2.3.1
Spark JobServer 0.8.0
Mesos 1.5.0
Spark cluster mode

For now, we turn back to use Spark standalone cluster mode. Everything is working now.

**Maybe an off topic question** I try to find some where some documentation of Spark job server about how CoarseGrainedExecutorBackend handles the application execution, but I could not find any.
My point is: SJS keeps an context running for a long time. The executor CoarseGrainedExecutorBackend keeps being alive during this duration. **Whether this long running executor can cause some potential memory leak (objects in memory of previous jobs might not be properly collected, ...)?**
Our usercase: We create a context. Using this context to run some jobs. We use just simple example LongPiJob application. After running 3 jobs of this application using the same context that we created. The RES memory increased from 0.2G -> 1.2G. 

I have checked the issue #557 which mentioned the same behavior. But at the end it does not seem to be answered. Should I open a dedicated issue for this question?
